arch/riscv/core/thread.c

```diff
@@ -196,6 +197,15 @@ int arch_float_enable(struct k_thread *thread, unsigned int options)
 }
 #endif /* CONFIG_FPU && CONFIG_FPU_SHARING */
 
+#ifdef CONFIG_THREAD_LOCAL_STORAGE
+void arch_switch_to_main_thread(struct k_thread *main_thread, char *stack_ptr,
+				k_thread_entry_t _main)
+{
+	__asm__("mv tp, %0" : : "r" (main_thread->tls));
+	z_swap_unlocked();
+}
+#endif
```

Comment:
Is this necessary?

Shouldn't `z_riscv_switch`, which was updated to load the TLS pointer into the `tp` register in the next commit ("riscv: Initialize TP register when starting threads"), be enough?

Reply:
I think so, but I hadn't tested that so I didn't elide this commit. I'd done all of my testing off the picolibc-module branch, I'm now re-doing that testing on the tls-fixes branch to make sure I've captured all of the necessary fixes. In particular, the stack smashing tests appear to mangle the TLS area when they overflow the stack.

Reply:
Just finished testing this -- the answer is 'no' -- the main thread is started without going through those paths, so we also need this patch to make sure TP is set everywhere.